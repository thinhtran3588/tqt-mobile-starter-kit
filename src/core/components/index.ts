export * from '@tqt/mobile/lib/components';
export * from './icon/icon.component';
export {
  Provider as PaperProvider,
  DefaultTheme as PaperDefaultTheme,
  Colors,
  Button,
  Appbar,
  Card,
  ActivityIndicator,
  Avatar,
  Badge,
} from 'react-native-paper';
export {StatusBar, Alert, ScrollView} from 'react-native';
