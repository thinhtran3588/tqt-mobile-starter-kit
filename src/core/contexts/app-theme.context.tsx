import React, {useContext, useEffect, useMemo} from 'react';
import {useImmer} from 'use-immer';
import {useColorScheme} from 'react-native';
import {usePersistenceImmer} from '@core/hooks';
import {Colors} from 'react-native-paper';

interface AppThemeProviderProps {
  children?: React.ReactNode;
}

interface AppThemeState {
  useSystemTheme: boolean;
  darkMode: boolean;
  theme: 'light' | 'dark';
  primaryColorId: string;
  primaryColor: string;
}

interface Dispatch {
  setUseSystemTheme: (useSystemTheme: boolean) => void;
  setDarkMode: (dark: boolean) => void;
  setPrimaryColor: (primaryColor: string) => void;
}

interface Color {
  id: string;
  color: string;
  darkColor: string;
  text: string;
}

export const COLORS_LOOKUP: {[id: string]: Color} = {
  AMBER: {
    id: 'AMBER',
    color: Colors.amber500,
    darkColor: Colors.amber300,
    text: 'Amber',
  },
  BLUE: {
    id: 'BLUE',
    color: Colors.blue500,
    darkColor: Colors.blue300,
    text: 'Blue',
  },
  BLUE_GREY: {
    id: 'BLUE_GREY',
    color: Colors.blueGrey500,
    darkColor: Colors.blueGrey300,
    text: 'Blue Grey',
  },
  BROWN: {
    id: 'BROWN',
    color: Colors.brown500,
    darkColor: Colors.brown300,
    text: 'Brown',
  },
  CYAN: {
    id: 'CYAN',
    color: Colors.cyan500,
    darkColor: Colors.cyan300,
    text: 'Cyan',
  },
  DEEP_ORANGE: {
    id: 'DEEP_ORANGE',
    color: Colors.deepOrange500,
    darkColor: Colors.deepOrange300,
    text: 'Deep Orange',
  },
  GREEN: {
    id: 'GREEN',
    color: Colors.green500,
    darkColor: Colors.green300,
    text: 'Green',
  },
  GREY: {
    id: 'GREY',
    color: Colors.grey500,
    darkColor: Colors.grey300,
    text: 'Grey',
  },
  INDIGO: {
    id: 'INDIGO',
    color: Colors.indigo500,
    darkColor: Colors.indigo300,
    text: 'Indigo',
  },
  LIGHT_BLUE: {
    id: 'LIGHT_BLUE',
    color: Colors.lightBlue500,
    darkColor: Colors.lightBlue300,
    text: 'Light Blue',
  },
  LIGHT_GREENS: {
    id: 'LIGHT_GREENS',
    color: Colors.lightGreen500,
    darkColor: Colors.lightGreen300,
    text: 'Light Green',
  },
  LIME: {
    id: 'LIME',
    color: Colors.lime500,
    darkColor: Colors.lime300,
    text: 'Lime',
  },
  ORANGE: {
    id: 'ORANGE',
    color: Colors.orange500,
    darkColor: Colors.orange300,
    text: 'Orange',
  },
  PINK: {
    id: 'PINK',
    color: Colors.pink500,
    darkColor: Colors.pink300,
    text: 'Pink',
  },
  PURPLE: {
    id: 'PURPLE',
    color: Colors.purple500,
    darkColor: Colors.purple300,
    text: 'Purple',
  },
  RED: {
    id: 'RED',
    color: Colors.red500,
    darkColor: Colors.red300,
    text: 'Red',
  },
  TEAL: {
    id: 'TEAL',
    color: Colors.teal500,
    darkColor: Colors.teal300,
    text: 'Teal',
  },
  YELLOW: {
    id: 'YELLOW',
    color: Colors.yellow500,
    darkColor: Colors.yellow300,
    text: 'Yellow',
  },
};
export const COLORS = Object.keys(COLORS_LOOKUP).map((key: string) => COLORS_LOOKUP[key]);

const DEFAULT_APP_THEME: AppThemeState = {
  useSystemTheme: true,
  darkMode: false,
  theme: 'light',
  primaryColorId: COLORS_LOOKUP.CYAN.id,
  primaryColor: COLORS_LOOKUP.CYAN.color,
};
export const LIGHT_BACKGROUND_COLOR = '#fff';
export const DARK_BACKGROUND_COLOR = '#272727';

const APP_THEME_KEY = 'APP_THEME';

const AppThemeContext = React.createContext(DEFAULT_APP_THEME);
const AppThemeDispatchContext = React.createContext<Dispatch>(undefined as never);

const AppThemeProvider = (props: AppThemeProviderProps): JSX.Element => {
  const {children} = props;
  const [appTheme, setAppTheme] = useImmer(DEFAULT_APP_THEME);
  const [setAppThemePersistence] = usePersistenceImmer(appTheme, setAppTheme, APP_THEME_KEY);
  const colorScheme = useColorScheme();

  const updateTheme = (draft: AppThemeState): void => {
    if (draft.useSystemTheme) {
      draft.theme = colorScheme === 'dark' ? 'dark' : 'light';
    } else {
      draft.theme = draft.darkMode ? 'dark' : 'light';
    }
  };

  useEffect(() => {
    setAppThemePersistence((draft) => {
      updateTheme(draft);
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [colorScheme]);

  const dispatch = useMemo(
    (): Dispatch => ({
      setUseSystemTheme: (useSystemTheme) => {
        setAppThemePersistence((draft) => {
          draft.useSystemTheme = useSystemTheme;
          updateTheme(draft);
        });
      },
      setDarkMode: (dark) => {
        setAppThemePersistence((draft) => {
          draft.darkMode = dark;
          updateTheme(draft);
        });
      },
      setPrimaryColor: (primaryColor) => {
        setAppThemePersistence((draft) => {
          draft.primaryColorId = primaryColor;
          updateTheme(draft);
        });
      },
    }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [],
  );
  return (
    <AppThemeContext.Provider value={appTheme}>
      <AppThemeDispatchContext.Provider value={dispatch}>{children}</AppThemeDispatchContext.Provider>
    </AppThemeContext.Provider>
  );
};

const useAppTheme = (): [AppThemeState, Dispatch] => {
  const theme = useContext(AppThemeContext);
  const dispatch = useContext(AppThemeDispatchContext);
  return [theme, dispatch];
};

export {AppThemeProvider, useAppTheme};
