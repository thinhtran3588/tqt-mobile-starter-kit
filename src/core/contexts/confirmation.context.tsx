import React, {useContext, useMemo} from 'react';
import {useImmer} from 'use-immer';
import {ColorType} from './app-theme.context';

interface ConfirmationProviderProps {
  children?: React.ReactNode;
}

export interface ConfirmationButton {
  text: string;
  onPress?: () => void;
  type?: ColorType;
}

export interface ConfirmationState {
  open?: boolean;
  message: string;
  buttons: ConfirmationButton[];
}

export interface OpenConfirmationParams {
  message: string;
  buttons: ConfirmationButton[];
}

interface Dispatch {
  openConfirmation: (params: OpenConfirmationParams) => void;
  closeConfirmation: () => void;
}

export const DEFAULT_CONFIRMATION: ConfirmationState = {
  open: false,
  message: '',
  buttons: [],
};

const ConfirmationContext = React.createContext(DEFAULT_CONFIRMATION);
const ConfirmationDispatchContext = React.createContext<Dispatch>(undefined as never);

const ConfirmationProvider = (props: ConfirmationProviderProps): JSX.Element => {
  const {children} = props;
  const [confirmation, setConfirmation] = useImmer(DEFAULT_CONFIRMATION);

  const dispatch = useMemo(
    (): Dispatch => ({
      openConfirmation: (params) => {
        setConfirmation((draft) => {
          draft.open = true;
          Object.assign(draft, params);
        });
      },
      closeConfirmation: () => {
        setConfirmation((draft) => Object.assign(draft, DEFAULT_CONFIRMATION));
      },
    }),
    [setConfirmation],
  );
  return (
    <ConfirmationContext.Provider value={confirmation}>
      <ConfirmationDispatchContext.Provider value={dispatch}>{children}</ConfirmationDispatchContext.Provider>
    </ConfirmationContext.Provider>
  );
};

const useConfirmation = (): {confirmation: ConfirmationState; dispatch: Dispatch} => {
  const confirmation = useContext(ConfirmationContext);
  const dispatch = useContext(ConfirmationDispatchContext);
  return {confirmation, dispatch};
};

export {ConfirmationProvider, useConfirmation};
