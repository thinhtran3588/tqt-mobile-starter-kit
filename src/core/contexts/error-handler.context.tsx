/* eslint-disable @typescript-eslint/no-explicit-any */
import React, {useEffect, useCallback} from 'react';
import {setJSExceptionHandler, setNativeExceptionHandler} from 'react-native-exception-handler';
import Promise from 'bluebird';
import {useTranslation} from 'react-i18next';
import {useNotification} from '@core/hooks/use-notification';
import {AppError, recordError} from '@core/exceptions';
import {useLoading} from './loading.context';

interface ErrorHandlerProviderProps {
  children?: React.ReactNode;
}

const ErrorHandlerContext = React.createContext(undefined);

const ErrorHandlerProvider = (props: ErrorHandlerProviderProps): JSX.Element => {
  const {children} = props;
  const {t} = useTranslation();
  const {showNotification} = useNotification();
  const {setLoading} = useLoading();

  const handleError = useCallback(
    (err: AppError): void => {
      const {code, messageCode, messageData} = err;

      showNotification({
        message: messageCode ? t(messageCode, messageData) : code || t('common:unknownError'),
        type: 'ERROR',
      });

      recordError(err);
    },
    [t, showNotification],
  );

  useEffect(() => {
    // For most use cases:
    // registering the error handler (maybe u can do this in the index.android.js or index.ios.js)
    setJSExceptionHandler((err: Error, _isFatal) => {
      handleError(err as AppError);
      // This is your custom global error handler
      // You do stuff like show an error dialog
      // or hit google analytics to track crashes
      // or hit a custom api to inform the dev team.
    }, true);

    // For most use cases:
    setNativeExceptionHandler((_exceptionString) => {
      // This is your custom global error handler
      // You do stuff like hit google analytics to track crashes.
      // or hit a custom api to inform the dev team.
      // NOTE: alert or showing any UI change via JS
      // WILL NOT WORK in case of NATIVE ERRORS.
    });

    // https://stackoverflow.com/questions/48487089/global-unhandledrejection-listener-in-react-native
    // We use the "Bluebird" lib for Promises, because it shows good perf
    // and it implements the "unhandledrejection" event:
    (global as any).Promise = Promise;

    // Global catch of unhandled Promise rejections:
    (global as any).onunhandledrejection = (err: AppError): void => {
      setLoading(false);
      // Warning: when running in "remote debug" mode (JS environment is Chrome browser),
      // this handler is called a second time by Bluebird with a custom "dom-event".
      // We need to filter this case out:
      handleError(err);
    };
  }, [handleError, setLoading]);

  return <ErrorHandlerContext.Provider value={undefined}>{children}</ErrorHandlerContext.Provider>;
};

export {ErrorHandlerProvider};
