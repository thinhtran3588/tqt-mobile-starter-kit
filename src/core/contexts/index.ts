export * from './app-theme.context';
