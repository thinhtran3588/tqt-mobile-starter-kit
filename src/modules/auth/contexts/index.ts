export * from './auth.context';
export * from './sign-in-toggle-clear-form.context';
