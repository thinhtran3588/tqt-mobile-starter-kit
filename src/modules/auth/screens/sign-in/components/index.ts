export * from './social-sign-in/social-sign-in.component';
