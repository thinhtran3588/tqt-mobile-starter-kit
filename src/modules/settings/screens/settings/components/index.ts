export * from './theme-setting/theme-setting.component';
export * from './primary-color-setting/primary-color-setting.component';
export * from './general-info/general-info.component';
export * from './language-setting/language-setting.component';
