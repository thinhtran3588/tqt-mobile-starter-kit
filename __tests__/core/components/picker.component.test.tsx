/* eslint-disable @typescript-eslint/no-explicit-any */
import React, {useState} from 'react';
import RNPicker from 'react-native-picker';
import {render, toJSON, fireEvent, act} from '@test-utils';
import {Picker, Colors, Button, PickerDataItem} from '@core/components';
import {COLORS} from '@core/contexts';

const colors: PickerDataItem[] = COLORS.map((c) => ({value: c.id, label: c.text}));

it('renders close correctly', async () => {
  const setPickerOpen = jest.fn();
  const setPrimaryColor = jest.fn();
  const {baseElement} = render(
    <Picker
      key='primary-color-picker'
      value={Colors.amber500}
      open={false}
      setOpen={setPickerOpen}
      dataSources={colors}
      onChangeValue={setPrimaryColor}
    />,
  );
  expect(toJSON(baseElement)).toMatchSnapshot();
});

it('renders open correctly', async () => {
  const setPickerOpen = jest.fn();
  const setPrimaryColor = jest.fn();
  const {baseElement} = render(
    <Picker
      key='primary-color-picker'
      value={Colors.amber500}
      open
      setOpen={setPickerOpen}
      dataSources={colors}
      onChangeValue={setPrimaryColor}
    />,
  );
  expect(toJSON(baseElement)).toMatchSnapshot();
});

it('renders with incorrect initial value correctly', async () => {
  const setPickerOpen = jest.fn();
  const setPrimaryColor = jest.fn();
  const {baseElement} = render(
    <Picker
      key='primary-color-picker'
      value=''
      open
      setOpen={setPickerOpen}
      dataSources={colors}
      onChangeValue={setPrimaryColor}
    />,
  );
  expect(toJSON(baseElement)).toMatchSnapshot();
});

it('opens picker and pick a value', async () => {
  const setPrimaryColor = jest.fn();
  const Wrapper = (): JSX.Element => {
    const ref = React.createRef();
    const [pickerOpen, setPickerOpen] = useState(false);
    const openPicker = (): void => {
      setPickerOpen(true);
      if (ref) {
        (ref.current as any).props.onShow();
      }
    };
    return (
      <>
        <Picker
          ref={ref}
          key='primary-color-picker'
          value={Colors.amber500}
          open={pickerOpen}
          setOpen={setPickerOpen}
          dataSources={colors}
          onChangeValue={setPrimaryColor}
        />
        <Button onPress={openPicker}>Open picker</Button>
      </>
    );
  };
  const {getByText} = render(<Wrapper />);
  fireEvent.press(getByText('Open picker'));
  expect(RNPicker.init).toHaveBeenCalled();
  expect(RNPicker.show).toHaveBeenCalled();

  act(() => {
    ((RNPicker.init as unknown) as jest.Mock).mock.calls[0][0].onPickerConfirm(['Cyan']);
  });
  expect(RNPicker.hide).toHaveBeenCalled();
  expect(setPrimaryColor).toHaveBeenCalledWith('CYAN');
});

it('opens picker and do nothing if pick the same value', async () => {
  const setPrimaryColor = jest.fn();
  const Wrapper = (): JSX.Element => {
    const ref = React.createRef();
    const [pickerOpen, setPickerOpen] = useState(false);
    const openPicker = (): void => {
      setPickerOpen(true);
      if (ref) {
        (ref.current as any).props.onShow();
      }
    };
    return (
      <>
        <Picker
          ref={ref}
          key='primary-color-picker'
          value={Colors.amber500}
          open={pickerOpen}
          setOpen={setPickerOpen}
          dataSources={colors}
          onChangeValue={setPrimaryColor}
        />
        <Button onPress={openPicker}>Open picker</Button>
      </>
    );
  };
  const {getByText} = render(<Wrapper />);
  fireEvent.press(getByText('Open picker'));
  expect(RNPicker.init).toHaveBeenCalled();
  expect(RNPicker.show).toHaveBeenCalled();

  act(() => {
    ((RNPicker.init as unknown) as jest.Mock).mock.calls[0][0].onPickerConfirm(['Amber']);
  });
  expect(RNPicker.hide).toHaveBeenCalled();
  expect(setPrimaryColor).not.toHaveBeenCalled();
});
