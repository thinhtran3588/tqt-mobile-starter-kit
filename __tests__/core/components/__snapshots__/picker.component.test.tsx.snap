// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders close correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Modal
        animated={true}
        hardwareAccelerated={false}
        supportedOrientations={
          Array [
            "portrait",
            "portrait-upside-down",
            "landscape",
            "landscape-left",
            "landscape-right",
          ]
        }
        transparent={true}
        visible={false}
      >
        <View
          accessible={true}
          focusable={true}
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <BlurView
            blurAmount={10}
            blurType="light"
            style={
              Array [
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                undefined,
              ]
            }
          />
        </View>
      </Modal>
    </View>
  </View>
</View>
`;

exports[`renders open correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Modal
        animated={true}
        hardwareAccelerated={false}
        supportedOrientations={
          Array [
            "portrait",
            "portrait-upside-down",
            "landscape",
            "landscape-left",
            "landscape-right",
          ]
        }
        transparent={true}
        visible={true}
      >
        <View
          accessible={true}
          focusable={true}
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <BlurView
            blurAmount={10}
            blurType="light"
            style={
              Array [
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                undefined,
              ]
            }
          />
        </View>
      </Modal>
    </View>
  </View>
</View>
`;

exports[`renders with incorrect initial value correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Modal
        animated={true}
        hardwareAccelerated={false}
        supportedOrientations={
          Array [
            "portrait",
            "portrait-upside-down",
            "landscape",
            "landscape-left",
            "landscape-right",
          ]
        }
        transparent={true}
        visible={true}
      >
        <View
          accessible={true}
          focusable={true}
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <BlurView
            blurAmount={10}
            blurType="light"
            style={
              Array [
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                undefined,
              ]
            }
          />
        </View>
      </Modal>
    </View>
  </View>
</View>
`;
